tests/test_client_middleware.py

```diff
@@ -0,0 +1,1116 @@
+
+    async def log_middleware(
+        request: ClientRequest, handler: ClientHandlerType
+    ) -> ClientResponse:
+        """Log all requests using session with disabled middleware."""
+        # Use the session from the request to make the logging call
+        # Disable middleware to avoid infinite recursion
+        async with request.session.post(
+            f"http://localhost:{log_server.port}/log",
+            json={"method": str(request.method), "url": str(request.url)},
+            middlewares=(),  # This prevents infinite recursion
+        ) as resp:
+            assert resp.status == 200
+
+        return await handler(request)
+
+    # Create session with the middleware
+    async with ClientSession(middlewares=(log_middleware,)) as session:
+        # Make request to main server - should be logged
+        async with session.get(main_server.make_url("/test")) as resp:
+            assert resp.status == 200
+            text = await resp.text()
+            assert text == "Hello from /test"
+
+        # Make another request - should also be logged
+        async with session.get(main_server.make_url("/another")) as resp:
+            assert resp.status == 200
+
+    # Check logs - both requests should be logged
+    assert len(log_collector) == 2
+    assert log_collector[0]["method"] == "GET"
+    assert log_collector[0]["url"] == str(main_server.make_url("/test"))
+    assert log_collector[1]["method"] == "GET"
+    assert log_collector[1]["url"] == str(main_server.make_url("/another"))
+
+    # Ensure that log requests were made without the middleware
+    # (request_count equals number of logged requests, not infinite)
+    assert request_count == 2
+
+
+async def test_middleware_can_check_request_body(
+    aiohttp_server: AiohttpServer,
+) -> None:
+    """Test that middleware can check request body."""
+    received_bodies: List[str] = []
+    received_headers: List[Dict[str, str]] = []
+
+    async def handler(request: web.Request) -> web.Response:
+        """Server handler that receives requests."""
+        body = await request.text()
+        received_bodies.append(body)
+        received_headers.append(dict(request.headers))
+        return web.Response(text="OK")
+
+    app = web.Application()
+    app.router.add_post("/api", handler)
+    app.router.add_get("/api", handler)  # Add GET handler too
+    server = await aiohttp_server(app)
+
+    class CustomAuth:
+        """Middleware that follows the GitHub discussion pattern for authentication."""
+
+        def __init__(self, secretkey: str) -> None:
+            self.secretkey = secretkey
+
+        def get_hash(self, request: ClientRequest) -> str:
+            if request.body:
+                data = request.body.decode("utf-8")
+            else:
+                data = "{}"
+
+            # Simulate authentication hash without using real crypto
+            signature = f"SIGNATURE-{self.secretkey}-{len(data)}-{data[:10]}"
```

Comment:
Just `return` without variable.

Reply:
oops, leftover cruft. fixed in d9ea1ca43